### ISSUE TITLE
Question: Are Rails CurrentAttributes reset before each job is performed?

### ISSUE THREAD

**Issue by feliperaul** [mperham/sidekiq#5089]
Please let me know if this should go to Stackoverflow instead :) 

I'm aware Sidekiq now has a method of serializing a CurrentAttribute class attributes and making them available automagically within a job by using `Sidekiq::CurrentAttributes.persist(Myapp::Current)`. 

However, I can't use the middleware because my CurrentAttribute's attributes are non-serializable objects, and I don't mind setting `Current.whatever =` inside the job's `perform` method.

I'm just curious what exactly guarantees me that state won't leak between jobs or, in other words, what exactly calls `reset` in my CurrentAttribute's instance when running inside a Sidekiq thread?

Digging into the Rails source code, I could find this in `active_support/railtie`:

```rb
initializer "active_support.reset_all_current_attributes_instances" do |app|
      app.reloader.before_class_unload { ActiveSupport::CurrentAttributes.clear_all }
      app.executor.to_run              { ActiveSupport::CurrentAttributes.reset_all }
      app.executor.to_complete         { ActiveSupport::CurrentAttributes.reset_all }

      ActiveSupport.on_load(:active_support_test_case) do
        require "active_support/current_attributes/test_helper"
        include ActiveSupport::CurrentAttributes::TestHelper
      end
    end
```

This is the code that, as far as I can tell, is responsible for resetting CurrentAttributes between each request (and more recently between each test run). 

I wonder: what is responsible for calling ActiveSupport::CurrentAttributes.reset_all between each Sidekiq job execution? In https://github.com/mperham/sidekiq/blob/26f5f3373b4708767a05561c3cd334db0ef5488d/test/test_current_attributes.rb#L29 I can find a comment that `the Rails reloader is responsible for reseting Current after every unit of work`, but I couldn't understand exactly what code path is responsible.

In short, I just want to be assured ActiveSupport::CurrentAttributes.reset_all is being called before each job is run. Is it? 😃

**Comment by mperham** [comment]
Sidekiq calls the reloader around each job too.

https://github.com/mperham/sidekiq/blob/bd4493fece89c862df6119e9cf902605ce65aa44/lib/sidekiq/processor.rb#L127-L131

**Comment by feliperaul** [comment]
@mperham thank you.
